gtfs/spec/en/reference.md

```diff
@@ -231,9 +243,11 @@ File: **Required**
 
 File: **Conditionally Required**
 
+Primary ID (`service_id`)
+
 |  Field Name | Type | Presence | Description |
 |  ------ | ------ | ------ |------ |
-|  `service_id` | ID | **Required** | Uniquely identifies a set of dates when service is available for one or more routes. Each `service_id` value can appear at most once in a [calendar.txt](#calendartxt) file. |
+|  `service_id` | Primary ID | **Required** | Uniquely identifies a set of dates when service is available for one or more routes. Each `service_id` value can appear at most once in a [calendar.txt](#calendartxt) file. |
```

Comment:
I disagree on this point. This is actually a foreign id pointing towards a not yet existing structure in GTFS: services. It is used as grouping. Technically one could argue that multiple ranges _could_ be supported.

Reply:
I agree that that would be a reasonable interpretation, but the spec already states that ```Each `service_id` value can appear at most once in a [calendar.txt](#calendartxt) file.```, which is essentially the definition of "Primary ID". I think changing that part of the spec would require a separate pull-request.

Reply:
It is still not a primary key. Because it does not have to be in the file, if only calendar_dates.txt is delivered...

Reply:
Sure, calendar.txt is **Conditionally Required**, but doesn't change the fact that `service_id` is **Required** if calendar.txt is provided, and is allowed to appear at most once in calendar.txt, and that's what makes it a primary key for the file.

Reply:
Conceptually there is a set of services which is the union of all service IDs referenced in calendar or calendar_dates. But that's just a set of IDs with no other attached fields, so it can be implicit. I believe the original idea was that all services would be defined in calendar.txt and exceptions referencing those services in calendar_dates.txt, but in some systems this would yield services in calendar.txt with no active days, and an informal rule was made that any such service with no active days can be dropped. For robustness and consistency, it might have been better to just require these empty services to be defined anyway in calendar.txt but that's not the spec we have. Looking at it this way, `service_id` is the primary key of `calendar`, and is a foreign key in `calendar_dates` which is allowed to reference these implicit fully-inactive rows in `calendar`.